### PR TITLE
executor: fix bug when correlated column is used in IndexMergeReader

### LINKED ISSUE #30913: Wrong result when using index-merge
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
/* test */ drop table if exists t1, t2;
/* test */ create table t1  (c_int int, c_str varchar(40), c_datetime datetime, c_decimal decimal(12, 6), primary key(c_int), key(c_int), key(c_str), unique key(c_decimal), key(c_datetime));
/* test */ create table t2  like t1 ;
/* test */ insert into t1 (c_int, c_str, c_datetime, c_decimal) values (6, 'sharp payne', '2020-06-07 10:40:39', 6.117000) , (7, 'objective kare', '2020-02-05 18:47:26', 1.053000) , (8, 'thirsty pasteur', '2020-01-02 13:06:56', 2.506000) , (9, 'blissful wilbur', '2020-06-04 11:34:04', 9.144000) , (10, 'reverent mclean', '2020-02-12 07:36:26', 7.751000) ;
/* test */ insert into t2 (c_int, c_str, c_datetime, c_decimal) values (6, 'beautiful joliot', '2020-01-16 01:44:37', 5.627000) , (7, 'hopeful blackburn', '2020-05-23 21:44:20', 7.890000) , (8, 'ecstatic davinci', '2020-02-01 12:27:17', 5.648000) , (9, 'hopeful lewin', '2020-05-05 05:58:25', 7.288000) , (10, 'sharp jennings', '2020-01-28 04:35:03', 9.758000) ;
/* test */ select * from t1 where c_decimal < all (select c_decimal from t2 where t1.c_int = t2.c_int and t1.c_datetime > t2.c_datetime and t2.c_decimal = 9.060 or t2.c_str <= 'interesting shtern' and t1.c_int = t2.c_int); -- flags: QUERY
```

### 2. What did you expect to see? (Required)

```sql
/* test */ select * from t1 where c_decimal < all (select c_decimal from t2 where t1.c_int = t2.c_int and t1.c_datetime > t2.c_datetime and t2.c_decimal = 9.060 or t2.c_str <= 'interesting shtern' and t1.c_int = t2.c_int); -- flags: QUERY
-- test >> +-------+-----------------+---------------------+-----------+
-- test    | c_int |      c_str      |     c_datetime      | c_decimal |
-- test    +-------+-----------------+---------------------+-----------+
-- test    |     7 | objective kare  | 2020-02-05 18:47:26 |  1.053000 |
-- test    |     8 | thirsty pasteur | 2020-01-02 13:06:56 |  2.506000 |
-- test    |    10 | reverent mclean | 2020-02-12 07:36:26 |  7.751000 |
-- test    +-------+-----------------+---------------------+-----------+
```

### 3. What did you see instead (Required)

```sql
/* test */ select * from t1 where c_decimal < all (select c_decimal from t2 where t1.c_int = t2.c_int and t1.c_datetime > t2.c_datetime and t2.c_decimal = 9.060 or t2.c_str <= 'interesting shtern' and t1.c_int = t2.c_int); -- flags: QUERY
-- test >> +-------+-----------------+---------------------+-----------+
-- test    | c_int |      c_str      |     c_datetime      | c_decimal |
-- test    +-------+-----------------+---------------------+-----------+
-- test    |     6 | sharp payne     | 2020-06-07 10:40:39 |  6.117000 |
-- test    |     7 | objective kare  | 2020-02-05 18:47:26 |  1.053000 |
-- test    |     8 | thirsty pasteur | 2020-01-02 13:06:56 |  2.506000 |
-- test    |     9 | blissful wilbur | 2020-06-04 11:34:04 |  9.144000 |
-- test    |    10 | reverent mclean | 2020-02-12 07:36:26 |  7.751000 |
-- test    +-------+-----------------+---------------------+-----------+
```

### 4. What is your TiDB version? (Required)

master @ f1e587704c0f8aa46e4c5bb33c390718e9e577f4

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #30913

Problem Summary: Checkout the following plan: we got `Apply` and correlated column in `Selection`, which is pushed to tikv. But `IndexMergeReader` doesn't handle correlated column at all. So the result is wrong.

    mysql> explain format=brief select * from t2 where c1 < all(select /*+ use_index_merge(t1) */ c1 from t1 where (c1 = 10 and c1 = t2.c3 or c2 = 1 and c2 = t2.c3));
    +--------------------------------+----------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
    | id                             | estRows  | task      | access object          | operator info                                                                                                                                           |
    +--------------------------------+----------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Projection                     | 10000.00 | root      |                        | test.t2.c1, test.t2.c2, test.t2.c3                                                                                                                      |
    | └─Apply                        | 10000.00 | root      |                        | CARTESIAN inner join, other cond:or(and(lt(test.t2.c1, Column#8), if(ne(Column#9, 0), NULL, 1)), or(eq(Column#10, 0), if(isnull(test.t2.c1), NULL, 0))) |
    |   ├─TableReader(Build)         | 10000.00 | root      |                        | data:TableFullScan                                                                                                                                      |
    |   │ └─TableFullScan            | 10000.00 | cop[tikv] | table:t2               | keep order:false, stats:pseudo                                                                                                                          |
    |   └─StreamAgg(Probe)           | 1.00     | root      |                        | funcs:min(test.t1.c1)->Column#8, funcs:sum(0)->Column#9, funcs:count(1)->Column#10                                                                      |
    |     └─IndexMerge               | 0.01     | root      |                        |                                                                                                                                                         |
    |       ├─Selection(Build)       | 1.00     | cop[tikv] |                        | eq(10, test.t2.c3)                                                                                                                                      |
    |       │ └─TableRangeScan       | 1.00     | cop[tikv] | table:t1               | range:[10,10], keep order:false, stats:pseudo                                                                                                           |
    |       ├─Selection(Build)       | 8.00     | cop[tikv] |                        | eq(1, test.t2.c3)                                                                                                                                       |
    |       │ └─IndexRangeScan       | 10.00    | cop[tikv] | table:t1, index:c2(c2) | range:[1,1], keep order:false, stats:pseudo                                                                                                             |
    |       └─Selection(Probe)       | 0.01     | cop[tikv] |                        | or(and(eq(test.t1.c1, 10), eq(10, test.t2.c3)), and(eq(test.t1.c2, 1), eq(1, test.t2.c3)))                                                              |
    |         └─TableRowIDScan       | 9.00     | cop[tikv] | table:t1               | keep order:false, stats:pseudo                                                                                                                          |
    +--------------------------------+----------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
    12 rows in set (0.00 sec)

### What is changed and how it works?
1. For correlated columns in filters, we will construct new dagPB.Executors each time.
  2.1 For each IndexScan in partialPlans, we will call `constructDistExec()` if necessary.
  2.2 For each TableScan in partialPlans, we will call `constructDistExec()` if necessary.
  2.3 For the final TableScan, we will call `constructDistExec()` if necessary.
2. For correlated columns in access conditions, we will rebuild range. The new kvRange will be generated for table/index scan.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
executor: fix bug when correlated column is used in IndexMergeReader
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

File: executor/index_merge_reader.go [hunk 1]
```diff
@@ -138,6 +149,33 @@ func (e *IndexMergeReaderExecutor) Open(ctx context.Context) (err error) {
...
 
+func (e *IndexMergeReaderExecutor) rebuildRangeForCorCol() (err error) {
+	len1 := len(e.partialPlans)
+	len2 := len(e.isCorColInPartialAccess)
+	if len1 != len2 {
+		return errors.Errorf("unexpect length for partialPlans(%d) and isCorColInPartialAccess(%d)", len1, len2)
+	}
+	for i, plan := range e.partialPlans {
+		if e.isCorColInPartialAccess[i] {
+			is, isIndexScan := plan[0].(*plannercore.PhysicalIndexScan)
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:160]: How about use switch to check type
```
	switch x := plan[0].(type) {
			case *plannercore.PhysicalTableScan:
				e.ranges[i], err = x.ResolveCorrelatedColumns()
			case *plannercore.PhysicalIndexScan:
				e.ranges[i], err = rebuildIndexRanges(e.ctx, x, x.IdxCols, x.IdxColLens)
			default:
				return errors.Errorf("unsupported plan type %T", plan[0])
			}
			if err != nil {
				return err
			}
```

File: executor/index_merge_reader_test.go [hunk 2]
```diff
@@ -454,3 +454,91 @@ func (test *testSerialSuite2) TestIndexMergeReaderMemTracker(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(bytes, Greater, 0.0)
 }
+
+func (test *testSerialSuite2) TestIndexMergeWithCorrelatedColumns(c *C) {
```
Comment (wshwsh12) [inline executor/index_merge_reader_test.go:458]: Should we add these into planner pkg?   see  planner/core/testdata

File: executor/index_merge_reader.go (hunk 1 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: done

File: executor/index_merge_reader_test.go (hunk 2 above)
Comment (guo-shaoge) [reply executor/index_merge_reader_test.go]: done

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**